pl-faded-parsons.js

```diff
@@ -1,377 +1,713 @@
+            });
+          } else {
+            alert("Your browser does not yet support this :(");
+          }
+        },
+      });
 
-        if (elem.style !== null) {
-          let raw_indent = parseInt(elem.style.marginLeft, 10);
-          code_indent = raw_indent / widget.options.x_indent;
+    $(this.config.toolbar)
+      .find(".widget-dark")
+      .on({ click: () => ParsonsGlobal.toggleDarkmode() });
+
+    const widget = this;
+    const solutionTray = $(widget.config.solutionList).sortable({
+      start: function (event, ui) {
+        $(ui.item).focus().addClass("codeline-in-motion");
+        widget.clearFeedback();
+      },
+      stop: function (event, ui) {
+        $(ui.item).blur().removeClass("codeline-in-motion");
+        if ($(event.target)[0] != ui.item.parent()[0]) {
+          return;
         }
 
-        this.code_indent = isNaN(code_indent) ? 0 : code_indent;
-    }
-    elem () { return this.elem; }
-    asText () {
-        let elemClone = $(this.elem).clone();
-        elemClone.find("input").each(function (_, inp) {
-            inp.replaceWith(inp.value);
-        });
-        elemClone[0].innerText = elemClone[0].innerText.trimRight();
-
-        let spaceCount = this.widget.options.x_indent * this.code_indent;
-        return " ".repeat(spaceCount) + elemClone[0].innerText;
-    }
-    asSegments () {
-        let elemClone = $(this.elem).clone();
-        let blank_values = []
-        elemClone.find("input").each(function (_, inp) {
-            blank_values.push(inp.value);
-            inp.replaceWith("!BLANK");
-        });
-        return {
-          given_segments : elemClone.text().split("!BLANK"),
-          blank_values   : blank_values
-        }
+        widget.updateIndent(
+          widget.calculateCodeIndent(
+            ui.position.left - ui.item.parent().position().left,
+            ui.item[0],
+          ),
+          ui.item[0],
+        );
+
+        widget.addLogEntry({ type: "moveOutput", target: ui.item[0].id }, true);
+      },
+      receive: function (_event, ui) {
+        widget.updateIndent(
+          widget.calculateCodeIndent(
+            ui.position.left - ui.item.parent().position().left,
+            ui.item[0],
+          ),
+          ui.item[0],
+        );
+
+        widget.addLogEntry({ type: "addOutput", target: ui.item[0].id }, true);
+      },
+      update: (e, ui) => widget.config.onSortableUpdate(e, ui),
+      grid: widget.config.canIndent ? [widget.config.xIndent, 1] : false,
+    });
+    solutionTray.addClass("output");
+
+    if (widget.config.starter) {
+      const starterTray = $(widget.config.starterList).sortable({
+        connectWith: solutionTray,
+        start: function (_event, ui) {
+          $(ui.item).focus().addClass("codeline-in-motion");
+          widget.clearFeedback();
+        },
+        receive: function (_event, ui) {
+          widget.updateIndent(0, ui.item[0].id);
```

Comment:
Quickfix: here should be `ui.item[0]`, this failure is silent to the user and logs a TypeError in console.